java/ql/lib/semmle/code/java/dataflow/internal/BaseSSA.qll

```diff
@@ -189,59 +190,46 @@ private module SsaInput implements SsaImplCommon::InputSig<Location, BasicBlock>
   }
 }
 
-private module Impl = SsaImplCommon::Make<Location, Cfg, SsaInput>;
+private module Impl = SsaImplCommon::Make<Location, Cfg, SsaImplInput>;
 
-private import Cached
+private module SsaInput implements Impl::SsaInputSig {
+  private import java as J
 
-cached
-private module Cached {
-  cached
-  VarRead getAUse(Impl::Definition def) {
-    BaseSsaStage::ref() and
-    exists(BaseSsaSourceVariable v, BasicBlock bb, int i |
-      Impl::ssaDefReachesRead(v, def, bb, i) and
-      result.getControlFlowNode() = bb.getNode(i) and
-      result = v.getAnAccess()
-    )
-  }
+  class Expr = J::Expr;
 
-  cached
-  predicate ssaDefReachesEndOfBlock(BasicBlock bb, Impl::Definition def) {
-    Impl::ssaDefReachesEndOfBlock(bb, def, _)
-  }
+  class Parameter = J::Parameter;
 
-  cached
-  predicate ssaUpdate(Impl::Definition def, VariableUpdate upd) {
-    exists(BaseSsaSourceVariable v, BasicBlock bb, int i |
-      def.definesAt(v, bb, i) and
-      variableUpdate(v, upd.getControlFlowNode(), bb, i) and
-      getDestVar(upd) = v
-    )
-  }
+  class VariableWrite = J::VariableWrite;
 
-  cached
-  predicate ssaImplicitInit(Impl::WriteDefinition def) {
-    exists(BaseSsaSourceVariable v, BasicBlock bb, int i |
-      def.definesAt(v, bb, i) and
-      hasEntryDef(v, bb) and
+  predicate explicitWrite(VariableWrite w, BasicBlock bb, int i, BaseSsaSourceVariable v) {
+    variableUpdate(v, w.asExpr().getControlFlowNode(), bb, i)
+    or
+    exists(Parameter p, Callable c |
+      c = p.getCallable() and
+      v = TLocalVar(c, p) and
+      w.isParameterInit(p) and
+      c.getBody().getBasicBlock() = bb and
       i = -1
     )
   }
+}
+
+module Ssa = Impl::MakeSsa<SsaInput>;
 
+import Ssa
+private import Cached
+
+cached
+private module Cached {
```

Comment:
Did you check that this cached stage is collapsed with the one inside `Impl::MakeSsa`?